awscli/autocomplete/indexer.py

```diff
@@ -0,0 +1,118 @@
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import logging
+import sqlite3
+
+
+LOG = logging.getLogger(__name__)
+
+
+class ModelIndexer(object):
+    def __init__(self, db_connection):
+        self._db_connection = db_connection
+
+    def generate_index(self, clidriver):
+        parent = 'aws'
+        self._db_connection.execute(
+            'INSERT OR REPLACE INTO command_table (command)'
+            'VALUES (:command)',
+            command=parent
+        )
+        command_table = clidriver.subcommand_table
+        self._generate_arg_index(command=parent, parent='',
+                                 arg_table=clidriver.arg_table)
+        self._generate_command_index(command_table, parent=parent)
+
+    def _generate_arg_index(self, command, parent, arg_table):
+        for name, value in arg_table.items():
+            self._db_connection.execute(
+                'INSERT INTO param_table '
+                '(argname, type_name, command, parent, nargs)'
+                ' VALUES (:argname, :type_name, :command, :parent, :nargs)',
+                argname=name, type_name=value.cli_type_name,
+                command=command, parent=parent,
+                nargs=value.nargs,
+            )
+
+    def _generate_command_index(self, command_table, parent):
+        for name, command in command_table.items():
+            self._db_connection.execute(
+                'INSERT INTO command_table (command, parent) '
+                'VALUES (:command, :parent)',
+                command=name, parent=parent,
+            )
+            self._generate_arg_index(command=name, parent=parent,
+                                     arg_table=command.arg_table)
+            self._generate_command_index(command.subcommand_table,
+                                         parent='%s.%s' % (parent, name))
+
+
+# This is similar to DBConnection in awscli.customization.history.
+# I'd like to reuse code, but we also have the contraint that we don't
+# want to import anything outside of awscli.autocomplete to ensure
+# our startup time is as minimal as possible.
+class DatabaseConnection(object):
+    _CREATE_CMD_TABLE = """\
+        CREATE TABLE IF NOT EXISTS command_table (
+          command TEXT,
+          parent TEXT REFERENCES command_table,
+          PRIMARY KEY (command, parent)
+        );
+    """
+    _CREATE_PARAM_TABLE = """\
+        CREATE TABLE IF NOT EXISTS param_table (
+          argname TEXT,
+          type_name TEXT,
+          command TEXT,
+          parent TEXT,
+          nargs TEXT,
```

Comment:
This will probably be addressed in future PR's but what will be the upgrading plan when we need to add new argument data to the table (i.e. say we wanted to add `required`, `choices`, or `group_name`)? In the history db, we used a generalized `payload` field so that if new history fields were added it would not force a complete regeneration of the history db. However, I do not think that is applicable to the completer db as we will probably want the entire db to be regenerated.

Reply:
Yeah it doesn't matter for auto completion.  The entire db is regenerated per release (the service models are going to be different).  This really only affects us (CLI devs) if we want to make changes to the schema while testing against a specific version of the CLI.